manuals/fvdeinfo.1
.Dd June 26, 2026
.Dt FVDEINFO 1
.Os
.Sh NAME
.Nm fvdeinfo
.Nd determines information about a Core Storage (CS) volume
.Sh SYNOPSIS
.Nm fvdeinfo
.Op Fl e Ar plist_path
.Op Fl k Ar key
.Op Fl o Ar offset
.Op Fl p Ar password
.Op Fl r Ar recovery_password
.Op Fl huvV
.Ar sources
.Sh DESCRIPTION
.Nm fvdeinfo
is a utility to determine information about a Core Storage (CS) volume
.Pp
.Nm fvdeinfo
is part of the
.Nm libfvde
package.
.Nm libfvde
is a library to access the Core Storage (CS) format
.Pp
.Ar sources
one or more source files or devices.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl e Ar plist_path
specify the path of the EncryptedRoot.plist.wipekey file
.It Fl h
shows this help
.It Fl k Ar key
specify the volume master key formatted in base16
.It Fl o Ar offset
specify the volume offset in bytes
.It Fl p Ar password
specify the password (or passphrase)
.It Fl r Ar recovery_password
specify the recovery password (or passphrase)
.It Fl u
unattended mode (disables user interaction)
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# fvdeinfo -o20480 image.raw
fvdeinfo 20220116
.sp
Core Storage information:
.sp
Logical volume group:
	Identifier			: 94923b58-9f31-4988-8707-cb90c8e45a46
	Name				: TESTLVG
	Number of physical volumes	: 1
	Number of logical volumes	: 1
.sp
Physical volume: 1
	Identifier			: 3273a055-3b8b-47e8-b970-df35eecda81b
	Size				: 511 MiB (536829952 bytes)
	Encryption method		: AES-XTS
.sp
Logical volume: 1
	Identifier			: 420af122-cf73-4a30-8b0a-a593a65fbef5
	Name				: TestLV
	Size				: 160 MiB (167772160 bytes)
	Is locked
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr fvdemount 1
.Sh AUTHORS
.An Omar Choudary <choudary.omar@gmail.com>, Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libfvde/issues
.Sh COPYRIGHT
Copyright (C) 2011-2026, Omar Choudary <choudary.omar@gmail.com>, \
Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* format support
  - handle volume offset detection of partial encrypted logical volume
  - use the VMK signature (a6 a6 a6 ...) to indicate the volume is unlocked not the unencrypted volume header?
  - volume with no 0x001a, 0x0305 and 0x0505 ?
  - remove libfvde_logical_volume_descriptor_get_last_block_number after debugging unsupported format variants?

* libfvde
  - libfvde_segment_descriptor_compare check for overlapping ranges
  - clean up: split off AES KEK functions
  - add handling of FVDE volumes with corrupt metadata, e.g. file header
  - store passwords as hashes?
  - read and compare secondary (physical) volume header
  - support out-of-order physical volume file IO pool

* add debug output
  - encryption method
  - checksum algorithm

* pyfvde
  - add functionality to set encrypted context plist from a file-like object

* fvdeinfo
  - print logical volume name and/or identifier when asking for a password

* fvdemount
  - print logical volume name and/or identifier when asking for a password

* code clean up
  - remove deprecated functions from libfvde and pyfvde
  - remove libfvde_volume_get_logical_volume_encryption_method
  - rename FileVault to CoreStorage ?
  - check for leading spaces
  - check for () without spaces
  - check for = alignment
  - check for if ( instead of if(
  - check error codes

* check README files
  - make sure pkg-config is mentioned as a dependency

20141018
* see `git log' for more recent change log
* removed README.macosx

20141017
* changes for deployment

20140907
* updated dependencies
* updated msvscpp files, not operational yet

20140531
* updated dependencies

20140112
* 2014 update
* updated dependencies
* exposed some encryption context plist functions in API

20130909
* updated dependencies
* worked on libcthreads build support

20131205
* updated dependencies

20130716
* updated dependencies
* replaced AES-XTS code by libcaes XTS implementation
* worked on git support
* code clean up

20130703
* code clean up

20130412
* updated README files
* updated dependencies
* added tests directory

20130331
* updated dependencies
* code clean up

20130305
* added safe-guards for XML plist data
* added check for block serial number
* added support for 0x0505 block that is defined before 0x001a block

20130228
* worked on improving logical volume offset detection

20130227
* worked on improving logical volume size detection
* code clean up

20130226
* worked on improving logical volume size detection
* code clean up

20130221
* fix for value overwrite of logical volume size by successive block

20130220
* small change to include/types.h

20130113
* 2013 update
* updated dependencies
* updated msvscpp files

20121217
* bug fixes

20121212
* updates to debug output

20121114
* updated libfvde.3 man page
* added volume get offset function
* worked on set keys
* code clean up
* worked on msvscpp files

20121113
* updated man pages
* added change to support number of iterations in password
  wrapped kek

20121112
* updated dependencies
* replaced libfvde_parse_uuid_string by libfguid function
* fixes in pc and spec file

20120916
* worked on fvdemount

20120915
* applied updates
* worked on fvdemount
  - added -X option
  - replaced EBADFD by EIO

20120813
* worked on encrypted removable media
* worked on man pages
* fixed issue for debug output for libfvde_io_handle_read_unencrypted_volume_header
* worked on function to pass VMK directly

20120808
* applied updates

20120724
* code clean up

20120723
* code clean up
* removed key and volume master key files they served no purpose in contrast
  to original libbde code

20120722
* code clean up
* fixed potential memory and data leak in set password functions

20120721
* fix for acinclude.m4 and libxml2.m4
* added libcsplit
* applied updates
* switched library order in tools
* updated .pc and .spec file
* preparations for msvscpp build

20120521
* configure.ac fixes for dpkg files

20120520
* code clean up

20120519
* removed libfvalue
* code clean up
  - renamed plist
  - change naming of volume offset and size
* add support for secondary encrypted metadata block
* added XML parsing to encrypted metadata block

20120518
* code clean up of fvdemount
* code clean up
  - removed more unused code
  - fix for relative volume offset

20120516
* bug fixes for plist
* code clean up
* added more debug output

20120514
* fixes for wide character build

20120510
* updated dependencies
* code clean up

20120405
* fix for libxml2 detection on MacOS-X

20120208
* code clean up

20120207
* code clean up
* fixed wrong context for decryption

20120206
* code clean up

20120202
* code clean up

20120124
* added debug output
* rewrite of dealing with metadata
* removed AES CCM code
* removed BDE metadata definitions

20120117
* added debug output

20120116
* 2012 update
* updated pc file
* updated dependencies
* removed libbde
* renamed to libfvde
* added libfcache
* updates for array type changes
* updates for libcaes changes
* updates for libbfio changes
* updated for libfcache/libfdata changes
* fixed a several potential memory leaks
* fixed a several integer overflow/truncation issues

20111007
* initial version based on libbde



fvdetools/fvdeinfo.c
/*
 * Shows information obtained from a Core Storage (CS) volume.
 *
 * Copyright (C) 2011-2026, Omar Choudary <choudary.omar@gmail.com>
 *                          Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "fvdetools_getopt.h"
#include "fvdetools_libcerror.h"
#include "fvdetools_libclocale.h"
#include "fvdetools_libcnotify.h"
#include "fvdetools_libfvde.h"
#include "fvdetools_output.h"
#include "fvdetools_signal.h"
#include "fvdetools_unused.h"
#include "info_handle.h"

info_handle_t *fvdeinfo_info_handle = NULL;
int fvdeinfo_abort                  = 0;

/* Signal handler for fvdeinfo
 */
void fvdeinfo_signal_handler(
      fvdetools_signal_t signal FVDETOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "fvdeinfo_signal_handler";

	FVDETOOLS_UNREFERENCED_PARAMETER( signal )

	fvdeinfo_abort = 1;

	if( fvdeinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     fvdeinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use fvdeinfo to determine information about a Core Storage (CS) volume.";

	fvdetools_option_t options[ ] = {
		{ 'e', "plist_path", "specify the path of the EncryptedRoot.plist.wipekey file" },
		{ 'h', NULL, "shows this help" },
		{ 'k', "key", "specify the volume master key formatted in base16" },
		{ 'o', "offset", "specify the volume offset in bytes" },
		{ 'p', "password", "specify the password (or passphrase)" },
		{ 'r', "recovery_password", "specify the recovery password (or passphrase)" },
		{ 'u', NULL, "unattended mode (disables user interaction)" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source volume" },
	};
	system_character_t options_string[ 32 ];

	libfvde_error_t *error                               = NULL;
	system_character_t * const *sources                  = NULL;
	system_character_t *option_encrypted_root_plist_path = NULL;
	system_character_t *option_key                       = NULL;
	system_character_t *option_password                  = NULL;
	system_character_t *option_recovery_password         = NULL;
	system_character_t *option_volume_offset             = NULL;
	char *program                                        = "fvdeinfo";
	system_integer_t option                              = 0;
	int number_of_options                                = (int) ( sizeof( options ) / sizeof( fvdetools_option_t ) );
	int number_of_sources                                = 0;
	int unattended_mode                                  = 0;
	int verbose                                          = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "fvdetools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( fvdetools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	fvdetools_output_version_fprint(
	 stdout,
	 program );

	if( fvdetools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = fvdetools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				fvdetools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'e':
				option_encrypted_root_plist_path = optarg;

				break;

			case (system_integer_t) 'h':
				fvdetools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'k':
				option_key = optarg;

				break;

			case (system_integer_t) 'o':
				option_volume_offset = optarg;

				break;

			case (system_integer_t) 'p':
				option_password = optarg;

				break;

			case (system_integer_t) 'r':
				option_recovery_password = optarg;

				break;

			case (system_integer_t) 'u':
				unattended_mode = 1;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				fvdetools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source files or devices.\n" );

		fvdetools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	sources           = &( argv[ optind ] );
	number_of_sources = argc - optind;

	libcnotify_verbose_set(
	 verbose );
	libfvde_notify_set_stream(
	 stderr,
	 NULL );
	libfvde_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &fvdeinfo_info_handle,
	     unattended_mode,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( option_encrypted_root_plist_path != NULL )
	{
		if( info_handle_set_encrypted_root_plist(
		     fvdeinfo_info_handle,
		     option_encrypted_root_plist_path,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set path of EncryptedRoot.plist.wipekey file.\n" );

			goto on_error;
		}
	}
	if( option_key != NULL )
	{
		if( info_handle_set_key(
		     fvdeinfo_info_handle,
		     option_key,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set key.\n" );

			goto on_error;
		}
	}
	if( option_password != NULL )
	{
		if( info_handle_set_password(
		     fvdeinfo_info_handle,
		     option_password,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set password.\n" );

			goto on_error;
		}
	}
	if( option_recovery_password != NULL )
	{
		if( info_handle_set_recovery_password(
		     fvdeinfo_info_handle,
		     option_recovery_password,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set recovery password.\n" );

			goto on_error;
		}
	}
	if( option_volume_offset != NULL )
	{
		if( info_handle_set_volume_offset(
		     fvdeinfo_info_handle,
		     option_volume_offset,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set volume offset.\n" );

			goto on_error;
		}
	}
	if( info_handle_open(
	     fvdeinfo_info_handle,
	     sources,
	     number_of_sources,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_SYSTEM ".\n",
		 sources[ 0 ] );

		goto on_error;
	}
	if( info_handle_volume_fprint(
	     fvdeinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print volume information.\n" );

		goto on_error;
	}
	if( info_handle_close(
	     fvdeinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &fvdeinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( fvdeinfo_info_handle != NULL )
	{
		info_handle_free(
		 &fvdeinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

